File: IOS/AttendanceCore/Sources/AttendanceCore/AttendanceRepository.swift
import Foundation

public final class AttendanceRepository: @unchecked Sendable {
    public static let defaultMinIntervalMillis: Int64 = 10 * 60 * 1000

    public enum RefreshResult: Equatable, Sendable {
        case success(AttendanceSnapshot, networkCallMade: Bool)
        case failure(MaitriError)
        case missingCredentials
    }

    private let credentialStore: CredentialStore
    private let snapshotStore: SnapshotStore
    private let client: MaitriClient
    private let minIntervalMillis: Int64
    private let now: () -> Int64

    public init(
        credentialStore: CredentialStore = CredentialStore(),
        snapshotStore: SnapshotStore = SnapshotStore(),
        client: MaitriClient = MaitriClient(),
        minIntervalMillis: Int64 = AttendanceRepository.defaultMinIntervalMillis,
        now: @escaping () -> Int64 = {
            Int64(Date().timeIntervalSince1970 * 1000)
        }
    ) {
        self.credentialStore = credentialStore
        self.snapshotStore = snapshotStore
        self.client = client
        self.minIntervalMillis = minIntervalMillis
        self.now = now
    }

    public var cachedSnapshot: AttendanceSnapshot? {
        snapshotStore.load()
    }

    public func refresh() async -> RefreshResult {
        if let cached = snapshotStore.load(),
           now() - cached.fetchedAtMillis < minIntervalMillis {
            return .success(cached, networkCallMade: false)
        }
        return await forceRefresh()
    }

    public func forceRefresh() async -> RefreshResult {
        guard let creds = credentialStore.load() else {
            return .missingCredentials
        }
        do {
            let fresh = try await client.fetchAttendance(creds: creds)
            try snapshotStore.save(fresh)
            return .success(fresh, networkCallMade: true)
        } catch let error as MaitriError {
            return .failure(error)
        } catch {
            return .failure(.portal(error.localizedDescription))
        }
    }

    public func testCredentials(_ creds: Credentials) async -> Result<AttendanceSnapshot, Error> {
        do {
            let snapshot = try await client.fetchAttendance(creds: creds)
            return .success(snapshot)
        } catch {
            return .failure(error)
        }
    }

    public func saveCredentials(_ creds: Credentials) throws {
        try credentialStore.save(creds)
    }

    public func clearCredentials() throws {
        try credentialStore.clear()
    }

    public var hasCredentials: Bool {
        credentialStore.hasCredentials
    }
}

File: Android/app/app/src/main/kotlin/edu/bmu/attendance/data/AttendanceRepository.kt
package edu.bmu.attendance.data

import android.content.Context

/**
 * Orchestrates [MaitriClient], [CredentialStore], [SnapshotStore], and [TimetableStore].
 *
 * Includes a small **rate limiter** (default 10 minutes): a refresh request
 * younger than that simply returns the cached snapshots. This keeps two
 * simultaneous triggers (e.g. WorkManager periodic + a user tap on Refresh)
 * from hammering Maitri, while still respecting the user's chosen 15-minute
 * widget refresh cadence.
 */
class AttendanceRepository(
    private val credentialStore: CredentialStore,
    private val snapshotStore: SnapshotStore,
    private val timetableStore: TimetableStore,
    private val client: MaitriClient = MaitriClient(),
    private val minIntervalMillis: Long = DEFAULT_MIN_INTERVAL_MILLIS,
    private val now: () -> Long = System::currentTimeMillis,
) {
    val cachedSnapshot: AttendanceSnapshot? get() = snapshotStore.load()
    val cachedTimetable: TimetableSnapshot? get() = timetableStore.load()

    /** Result of a [refresh] / [forceRefresh] call. */
    sealed class RefreshResult {
        data class Success(
            val snapshot: AttendanceSnapshot,
            val timetable: TimetableSnapshot?,
            val networkCallMade: Boolean,
        ) : RefreshResult()
        data class Failure(val error: MaitriError) : RefreshResult()
        object MissingCredentials : RefreshResult()
    }

    /**
     * Fetch fresh attendance + timetable unless the attendance cache is younger
     * than [minIntervalMillis], in which case both caches are returned as-is.
     */
    suspend fun refresh(): RefreshResult {
        val cached = snapshotStore.load()
        if (cached != null && now() - cached.fetchedAtMillis < minIntervalMillis) {
            return RefreshResult.Success(
                snapshot = cached,
                timetable = timetableStore.load(),
                networkCallMade = false,
            )
        }
        return forceRefresh()
    }

    /** Always hit the network (one login for attendance + current-week timetable). */
    suspend fun forceRefresh(): RefreshResult {
        val creds = credentialStore.load() ?: return RefreshResult.MissingCredentials
        return try {
            val (attendance, timetable) = client.fetchAttendanceAndTimetable(creds)
            snapshotStore.save(attendance)
            if (timetable != null) {
                timetableStore.save(timetable)
            }
            RefreshResult.Success(
                snapshot = attendance,
                timetable = timetable ?: timetableStore.load(),
                networkCallMade = true,
            )
        } catch (e: MaitriError) {
            RefreshResult.Failure(e)
        } catch (e: Exception) {
            RefreshResult.Failure(MaitriError.Portal(e.message ?: "Unexpected error"))
        }
    }

    /**
     * Test the supplied credentials without persisting them or affecting the
     * cached snapshot. Used by the Settings "Test login" button.
     */
    suspend fun testCredentials(creds: Credentials): Result<AttendanceSnapshot> = runCatching {
        client.fetchAttendance(creds)
    }

    fun clearCaches() {
        snapshotStore.clear()
        timetableStore.clear()
    }

    companion object {
        const val DEFAULT_MIN_INTERVAL_MILLIS: Long = 10 * 60 * 1000L

        /** App-wide singleton. Cheap to construct, but reused across UI/widget/worker. */
        @Volatile private var instance: AttendanceRepository? = null

        fun get(context: Context): AttendanceRepository {
            return instance ?: synchronized(this) {
                instance ?: AttendanceRepository(
                    credentialStore = CredentialStore(context),
                    snapshotStore = SnapshotStore(context),
                    timetableStore = TimetableStore(context),
                ).also { instance = it }
            }
        }
    }
}
